xscen/ensembles.py

```diff
@@ -441,7 +475,85 @@ def generate_weights(
+                (info[k][att] is None or len(info[k][att]) == 0) for k in info
+            ):
+                raise ValueError(
+                    f"The {att} attribute is missing from some simulations."
+                )
+            # Split dict and xr.Dataarray weights
+            if isinstance(v_att, xr.DataArray):
+                non_stationary_weights[att] = v_att
+            elif isinstance(v_att, dict):
+                stationary_weights[att] = v_att
+            else:
+                raise ValueError("Attribute_weights should be dict or xr.DataArray.")
+        # Stationary weights (dicts)
+        if stationary_weights:
+            for att, w_dict in stationary_weights.items():
+                for k, v in info.items():
+                    if v[att] not in w_dict and "others" not in w_dict:
+                        raise ValueError(
+                            f"The {att} {v[att]} or others are not in the attribute_weights dict."
+                        )
+                    elif v[att] not in w_dict and "others" in w_dict:
+                        w = w_dict["others"]
+                    elif v[att] in w_dict:
+                        w = w_dict[v[att]]
+                    weights.loc[{"realization": k}] = weights.sel(realization=k) * w
+        # Non-staionary weights (xr.DataArray)
+        if non_stationary_weights:
+            for att, da in non_stationary_weights.items():
+                # verification
+                if att not in da.coords:
+                    raise ValueError(f"{att} is not in the xr.DataArray coords.")
+                ls_coord = list(da.coords)
+                ls_coord.remove(att)
+                if len(ls_coord) > 1:
+                    raise ValueError(
+                        f"The {att} DataArray has more than one coord dimension to apply weights."
+                    )
+                else:
+                    coord = ls_coord[0]
+                # coord which will be used for broadcasting
+                if coord not in weights.coords:
+                    weights = weights.expand_dims({coord: da[coord].values})
+                ls_da = []
+                for k, v in info.items():
+                    if v[att] not in da[att] and "others" not in da[att]:
+                        raise ValueError(
+                            f"The {att} {v[att]} or others are not in the attribute_weights datarray coords."
+                        )
+                    elif v[att] not in da[att] and "others" in da[att]:
+                        ls_da.append(da.sel(**{att: "others"}).drop_vars(att))
+                    else:
+                        ls_da.append(da.sel(**{att: v[att]}).drop_vars(att))
+                nw = xr.concat(ls_da, weights.realization)
+                weights = weights * nw
+
     if standardize:
-        weights = weights / weights.sum(dim="realization")
+        if attribute_weights:
+            weights = weights / weights.sum(dim=weights.dims)
```

Comment:
Linked to the previous comment. If we get rid of the extra dimensions created for non-stationary weights, then we can get rid of this line.

At minimum, we should exclude `time/horizon` from here.